pass_args_to_sdf.md

```diff
@@ -0,0 +1,201 @@
+# Pass arguments to SDF files
+
+## Overview
+
+The goal of passing arguments to SDF files is to allow a user to send custom data into a model or world file and prevent model/world file duplication. Currently, there are multiple robot models with the same base but different sensor configurations. For example, in [SubT's robots](https://github.com/osrf/subt/wiki/Robots), X1 has 8-10 different sensor configurations but all have the same base. If we take a look at all the different configurations in [Fuel](https://app.ignitionrobotics.org/), we see that they are essentially copies of each other but with different sensors or different parameters for these sensors. The aim of adding parameter support is to avoid this duplication, where a model/world file contains all configurations with default parameters and a user may update and/or override certain parameters as desired.
+
+### Features
+
+The list of intended **primary features** include:
+* Sending data from a world or launch file into a model (e.g., topic names, change parameter values in model(s))
+* Enable/disable parts of models and/or worlds
+
+**Secondary features** include:
+* Add support to [ign-fuel-tools](https://github.com/ignitionrobotics/ign-fuel-tools) and [Fuel](https://app.ignitionrobotics.org/) web application
+
+### Need to consider:
+
+* Do other libraries (e.g., [ign-physics](https://github.com/ignitionrobotics/ign-physics)) need this capability?
+
+## Implementation
+
+The initial question was whether this process should be a [SDF](http://sdformat.org/) or [ign-gazebo](https://github.com/ignitionrobotics/ign-gazebo) feature. The conclusion is that this will most likely be implemented in ign-gazebo. The reasoning behind this is other organizations use SDF with other simulators, which raises concerns that it could affect their work. Also, this desired argument passing process is intended for ign-gazebo use. Although there may be minor updates to SDFormat and/or libsdformat, the current plan is to implement the main functionality in ign-gazebo. Below describes the current idea to design this new feature.
```

Comment:
> The reasoning behind this is other organizations use SDF with other simulators, which raises concerns that it could affect their work.

I understand the concern, writing something generic is always more complicated than writing something specific. But in this case, I think that functionality like adding / removing / modifying included elements could be considered general enough to be supported directly into SDFormat.

The part about using ERB to set parameters, on the other hand, would be too disruptive to add to SDF, introducing a large, and often complicated dependency (Ruby). So I agree that this part should be implemented at the Gazebo / Launch level.

Reply:
What do you think @scpeters ?

Reply:
anything we add to the SDFormat specification increases its complexity, so we have to be judicious about what becomes part of the official spec. I think it is possible that this type of functionality eventually makes its way into the official spec, but I think we should start by using custom elements parsed by `ign-gazebo` as you propose here (it could be elegant to create a plugin system for libsdformat that would parse custom elements, but that would take some additional design work). If after demonstrating the feature in `ign-gazebo`, if there is broader interest in the feature, we can propose adding it to the official spec.

I also agree that `ERB` support shouldn't be part of the official spec

Reply:
FWIW, Drake has a very similar pain point with SDFormat, and we would be eager to take inspiration from whatever solutions (or specification changes) you come up with, albeit at our usual somewhat slower pace.

Reply:
> Some concerns I have with making this `ign-gazebo` specific at first:

those concerns make sense to me. In that case, I would target some of these extensions of the `<include>` element at the SDFormat 1.8 composition proposal:

* http://sdformat.org/tutorials?tut=composition_proposal&cat=pose_semantics_docs&
* https://github.com/osrf/sdf_tutorials/blob/master/composition/proposal.md